test/clj/rems/db/test_transactions.clj

```diff
@@ -0,0 +1,177 @@
+        form-id (:id (form/create-form! user-id
+                                        {:organization ""
+                                         :title ""
+                                         :fields []}))
+        res-id (:id (resource/create-resource! {:resid (str "urn:uuid:" (UUID/randomUUID))
+                                                :organization ""
+                                                :licenses []}
+                                               user-id))
+        cat-id (:id (catalogue/create-catalogue-item! {:title ""
+                                                       :form form-id
+                                                       :resid res-id
+                                                       :wfid workflow-id}))
+        app-id (:application-id (applications/create-application! user-id [cat-id]))]
+    (assert app-id)
+    app-id))
+
+(defn- transaction-conflict? [^Exception e]
+  (cond
+    (nil? e) false
+    (instance? PSQLException e) (.contains (.getMessage e)
+                                           "The transaction might succeed if retried")
+    :else (transaction-conflict? (.getCause e))))
+
+(defn- sample-until-interrupted [f]
+  (loop [results []]
+    (if (.isInterrupted (Thread/currentThread))
+      results
+      (recur (try
+               (conj results (f))
+               (catch InterruptedException _
+                 (.interrupt (Thread/currentThread))
+                 results)
+               ;; XXX: HikariPool.getConnection wraps InterruptedException into SQLException
+               (catch SQLException e
+                 (if (instance? InterruptedException (.getCause e))
+                   (do (.interrupt (Thread/currentThread))
+                       results)
+                   (throw e))))))))
+
+(defn- submit-all [^ExecutorService thread-pool tasks]
+  (doall (for [task tasks]
+           (.submit thread-pool ^Callable task))))
+
+(deftest test-event-publishing-consistency
+  (let [test-duration-millis 2000
+        applications-count 5
+        writers-per-application 5
+        concurrent-readers 5
+        user-id (create-dummy-user)
+        app-ids (vec (for [_ (range applications-count)]
+                       (create-dummy-application user-id)))
+        observed-app-version-marker 999
+        mark-observed-app-version (fn [result _cmd application]
+                                    (if (and (:success result)
+                                             (= :application.event/draft-saved (:event/type (:result result))))
+                                      (assoc-in result [:result :application/field-values observed-app-version-marker]
+                                                (str (count (:application/events application))))
+                                      result))
+        write-event (fn [app-id]
+                      (try
+                        ;; Note: currently these tests pass only with serializable isolation
```

Comment:
and thus that's what we use in the middleware? please clarify

Reply:
I'll first try adding a `select for update` which should reduce transaction conflicts and also make this work at any isolation level.